### ISSUE TITLE
SmartCase interaction with advanced PerlRE escapes

### ISSUE THREAD

**Issue by n1vux** [beyondgrep/ack3#187]
a gotcha I've noticed with SmartCase ...
use of `\K` as cut/keep turns off smart case.

For 3.00, we should add that caveat in Cookbook where I recommend `\K` and also in Manual / HELP with SmartCase, that an explicit `-i` or `(?i-:)` is needed when using advanced PerlRE escapes such as `\K` `\Q` etc.

(This is because we're quite reasonably inspecting the string of the pattern RE and not inspecting the internal of the compiled QR, which is what we'd have to do to (a) avoid the rude message with -w and punctuation and DWIM, or to pretend `\K`  isn't a capital letter in the string and leave SmartCase on. Either that or find use the PP equivalent for PerlREs ... )

**Comment by petdance** [comment]
See also #156

**Comment by n1vux** [comment]
I don't know if this is reasonable or possible to fix. 
The SmartCase DWIM does the simple thing, and `'\K'` and `'\S'` do indeed contain capital letters.

In order to avoid parsing quotes ourselves, we'd need a Devel::??? XS module that walks internal form of an RE and lets us inspect tagged tokens. 
(`use re qw(debug DUMP)` gives output that shows the parse, but it's only output. The functions exported don't return all literals, only e.g. `regmust`, longest anchored and longest floating.)
(It's unclear to me how `B::REGEXP`'s minimal SV access methods would be much use to us.)

May be worth emphasizing in docs that those of us with Smartcase in our ACKRC need explicit `-i` when using uppercase escapes.

**Comment by petdance** [comment]
> I don't know if this is reasonable or possible to fix.

I know we've talked about it before, and it feels like it's severe enough that we seriously need to figure out a way.

It is very surprising to have this be case-insensitive

* `select . from table`

but this is case-sensitive

* `select \S+ from table`

but you can get around it with this

* `select [^\s]+ from table`

**Comment by n1vux** [comment]
On Sat, Jun 29, 2019, 10:41 AM Andy Lester <notifications@github.com> wrote:

> I don't know if this is reasonable or possible to fix.
>
> I know we've talked about it before, and it feels like it's severe enough
> that we seriously need to figure out a way.
>
> It is very surprising
>

I am in agreement it would be good to fix if fixable!

General rule is avoiding parsing perl with RE, and parsing an RE with RE
falls under that. I would like to avoid parsing RE escapes in Perl, too
many edgecases to do 100% job.
Hence my bias for doc fix before code fix.
But doing a 95% instead of 70% may be good enough.

One advantage here is if we extract the pattern string from the compiled
qr// either as a coercion  (interpolation into qq () or Regexp::Util), it
will have been canonicalized, it won't have any unnecessary escapes to
confuse a naive parse.

If we only do the advanced check for unescaped Caps only  if the simple
check for any Caps triggers, it won't take hardly any time (one extra if)
in usual case.

Is it easier than that?
Can we simply check, if Smartcase is on,
 ```
$re=qr/$pat/oi;
$pat="$re"; # canonicalized
$re=qr/$pat/o if  $pat =~ /(?:^|[^\\])[A-Z]/;
```

(OR better named unicode class for upper, `\p{XPosixUpper}` so that
accented cap E will turn off smart case too?)

**Comment by petdance** [comment]
I've started a branch `smartcase` for experimenting with.  I've wrapped up the lowercase detection in `App::Ack::is_lowercase` and created `t/lowercase.t`.  I've also started adding test cases that fail.

We've talked about this a bunch.  Now I want to start writing code and seeing what we can do, and how we can break our detection.

**Comment by petdance** [comment]
Reminder: You can't use `\Q` or `\E` in Perl regexes.  Those are string escapes.

**Comment by petdance** [comment]
@n1vux Take a look at the `smartcase` branch.  I took a low-tech way at attacking the problem.  I just ripped metacharacters out of the pattern and looked at what was left.  I'm pretty confident in my approach.  It's not parsing.  It's just mangling the string.

Please take a look at it and see if you can trick it.

**Comment by petdance** [comment]
Fixed at 440118fd3ee684f2860d3e0010e1528eb391cb5e.  Will be in v3.0.2.
